sim/core/item_swaps.go

```diff
@@ -0,0 +1,247 @@
+	})
+}
+
+func (swap *ItemSwap) GetItem(slot proto.ItemSlot) Item {
+	if slot-offset < 0 {
+		panic("Not able to swap Item " + slot.String() + " not supported")
+	}
+	return swap.unEquippedItems[slot-offset]
+}
+
+func (swap *ItemSwap) setItem(slot proto.ItemSlot, item Item) {
+	swap.unEquippedItems[slot-offset] = item
+}
+
+func (swap *ItemSwap) SwapItems(sim *Simulation, slots []proto.ItemSlot, useGCD bool) {
+	if !swap.IsEnabled() {
+		return
+	}
+
+	character := swap.character
+
+	meeleWeaponSwapped := false
+	newStats := stats.Stats{}
+	has2H := swap.GetItem(proto.ItemSlot_ItemSlotMainHand).HandType == proto.HandType_HandTypeTwoHand
+	for _, slot := range slots {
+		//will swap both on the MainHand Slot for 2H.
+		if slot == proto.ItemSlot_ItemSlotOffHand && has2H {
+			continue
+		}
+
+		if ok, swapStats := swap.swapItem(sim, slot, has2H); ok {
+			newStats = newStats.Add(swapStats)
+			meeleWeaponSwapped = slot == proto.ItemSlot_ItemSlotMainHand || slot == proto.ItemSlot_ItemSlotOffHand
+		}
+	}
+
+	character.AddStatsDynamic(sim, newStats)
+
+	for _, onSwap := range swap.onSwapCallbacks {
+		onSwap(sim)
+	}
+
+	if character.AutoAttacks.IsEnabled() && meeleWeaponSwapped {
+		character.AutoAttacks.StopMeleeUntil(sim, sim.CurrentTime, false)
+	}
+
+	if useGCD {
+		character.GCD.Set(1500 * time.Millisecond)
+	}
+}
+
+func (swap *ItemSwap) swapItem(sim *Simulation, slot proto.ItemSlot, has2H bool) (bool, stats.Stats) {
+	character := swap.character
+	oldItem := character.Equip[slot]
+	newItem := swap.GetItem(slot)
+
+	if newItem.ID == 0 && !(has2H && slot == proto.ItemSlot_ItemSlotOffHand) {
+		return false, stats.Stats{}
+	}
+	character.Equip[slot] = newItem
+	newStats := newItem.Stats.Add(oldItem.Stats.Multiply(-1))
+
+	//2H will swap out the offhand also.
+	if has2H && slot == proto.ItemSlot_ItemSlotMainHand {
+		_, ohStats := swap.swapItem(sim, proto.ItemSlot_ItemSlotOffHand, has2H)
+		newStats = newStats.Add(ohStats)
+	}
+
+	swap.setItem(slot, oldItem)
+	swap.swapWeapon(sim, slot)
+
+	return true, newStats
+}
+
+func (swap *ItemSwap) swapWeapon(sim *Simulation, slot proto.ItemSlot) {
+	character := swap.character
+	if !character.AutoAttacks.IsEnabled() {
+		return
+	}
+
+	switch slot {
+	case proto.ItemSlot_ItemSlotMainHand:
+		character.AutoAttacks.MH = character.WeaponFromMainHand(swap.mhCritMultiplier)
+		break
+	case proto.ItemSlot_ItemSlotOffHand:
+		character.AutoAttacks.OH = character.WeaponFromOffHand(swap.ohCritMultiplier)
+		//Special case for when the OHAuto Spell was setup with a non weapon and does not have a crit multiplier.
+		character.AutoAttacks.OHAuto.CritMultiplier = swap.ohCritMultiplier
+		break
+	case proto.ItemSlot_ItemSlotRanged:
+		character.AutoAttacks.Ranged = character.WeaponFromRanged(swap.rangedCritMultiplier)
+		break
+	}
+
+	character.AutoAttacks.IsDualWielding = character.Equip[proto.ItemSlot_ItemSlotMainHand].SwingSpeed != 0 && character.Equip[proto.ItemSlot_ItemSlotOffHand].SwingSpeed != 0
```

Comment:
I don't see any shield support, but you'll want to also set `character.PseudoStats.CanBlock` here.